lib/ruby_lsp/indexer.rb

```diff
@@ -0,0 +1,68 @@
+# typed: true
+# frozen_string_literal: true
+
+class Indexer
+  VM_DEFINECLASS_TYPE_CLASS           = 0x00
+  VM_DEFINECLASS_TYPE_SINGLETON_CLASS = 0x01
+  VM_DEFINECLASS_TYPE_MODULE          = 0x02
+  VM_DEFINECLASS_FLAG_SCOPED          = 0x08
+  VM_DEFINECLASS_FLAG_HAS_SUPERCLASS  = 0x10
+
+  def initialize(root)
+    @root = root
+    @index = {
+      constant: (Hash.new { |h,k| h[k] = [] }),
+      method: (Hash.new { |h,k| h[k] = [] })
+    }
+  end
+
+  def run
+    Dir["#{@root}/**/*.rb"].each do |file|
+      index_file(file)
+    end
+  end
+
+  def locs_for_symbol(symbol, kind)
+    $stderr.puts "--- Looking for #{symbol} of kind #{kind}"
+    index.dig(kind, symbol)
+  end
+
+  private
+
+  attr_reader :root, :index
+
+  def index_file(file)
+    iseq = RubyVM::InstructionSequence.compile(File.read(file)).to_a
```

Comment:
Does compiling the instruction sequences include comments? Indexing will also unlock properly implementing hover, but we'll need to keep track of the comments associated to each method/constant.

Reply:
In this case, we'd be compiling the file into instruction sequences and then parsing it with Ripper for the comments. Should we merge both things and just create a Ripper subclass that extracts only constant, methods and their comments?

Reply:
You could if you wanted, but I wouldn't. 

The issue is, if you're using ripper to get all of the constants and methods, then you have to deal with every variation in syntax. You need to define handlers for every child node that _could_ be in the tree. And you need to allocate a ton of objects. You basically have to manually build your syntax tree and maintain it. At that point you may as well use Syntax Tree.

On the other hand, if you're just lexing the file to get out the comments, it's going to be much faster. And it won't allocate nearly as many objects.

Reply:
We could also subclass `SyntaxTree::Parser` to intercept only the methods we're interested in. Not sure which would be faster (compiling inseq + parsing comments vs doing it in one go with a subclass). Whatever the fastest option is, since indexing is expensive.